Updates for demo

jnu opened the pull request:
- schema updates for program/dataset

Still to do:
- [x] clean up some duplication and nested maps

jnu committed:
small schema updates to dataset queries

jnu committed:
set up more mock data

jnu committed:
useState example

Commit:
feat(ch129): add queries and mock data for mock server
- update local schema json
- add datasets to mock gql query resolver'
- generate types from gql schema for  import

Commit:
refactor(ch129): create mock env var for switching host server uri
- added a script to run app with a mock env variable
- changed apollo client uri to conditional for mock env

Commit:
style(ch129): fix linting

Commit:
feat(ch129): add gql queries to component pages and update edit route

seerocode commented on client/src/components/AppSidebar.tsx:
```diff
@@ -37,6 +29,26 @@ const programs: any = [
+    { variables: { id: "1" } }
+  );
+
+  const datasets = data?.user?.teams.map((team) =>
```
This isn't being used right now and needs to be fixed for the sidebar.

seerocode replied:
Removed.

seerocode commented on client/src/components/DatasetRecordsTable.tsx:
```diff
@@ -1,183 +1,119 @@
-      men: 1,
-      women: 2,
-      nonBinary: 0,
+  const columns = [
```
I still DO NOT like that these columns aren't generated dynamically but we'll get there!

jnu replied:
agree! totally fine to leave this for now!

seerocode commented on client/src/components/DatasetRecordsTable.tsx:
```diff
@@ -1,183 +1,119 @@
+          <Space>
+            <Link
+              to={{
+                pathname: `/dataset/${datasetId}/entry/edit/${recordId}`,
```
How do we feel about this path for editing? Not sure it's necessary.

seerocode commented on client/src/index.tsx:
```diff
@@ -12,7 +12,10 @@ import {
-  uri: "/graphql/",
+  uri:
+    process.env.REACT_APP_ENV === "mock"
+      ? "http://localhost:4000"
```
Could this host go in the env file too? I tried for half a second and abandoned ship.

jnu replied:
this is totally fine, env files i actually find a little unnecessarily difficult to work with

Commit:
feat(ch129): Update gql schema

Commit:
test(ch129): update snapshot test

seerocode commented on api/schema.graphql:
```diff
@@ -74,6 +75,11 @@ type Record {
+  data: [Data!]!
+}
+
+type Data {
```
Separated category information into a Data type.
CC @nmanduj1 for review on this change

Commit:
test(ch129): update failing test
- Replace enzyme adapter while react 17 adapter is being released
- Wrap sidebar component with browserrouter
- Update package for enzyme adapter

Commit:
fix(ch129): change query imports to apollo client library

Commit:
refactor(ch129): make data input form dynamic

seerocode commented:
Moving tests and PR for finishing ch129 to a new PR since this one is getting bloated.

- [ ] mutations for add/edit page to update information on the details page
- [ ] tests!!!!
- [ ] hook up gql data to add/edit page

jnu commented on client/src/components/AppSidebar.tsx:
```diff
@@ -5,29 +5,21 @@ import { TeamOutlined, BarChartOutlined } from "@ant-design/icons";
 
 // TODO: add gql query for retrieving datasets a user has access to
-const programs: any = [
+const programs = [
```
just curious, why do we need this list of programs since we have the mock-server to fetch from?

seerocode replied:
There's an issue with how the data is getting mapped with this code:

```
const sidebarPrograms = data?.user?.teams.map((team) =>
    team.programs.map((program) => {
      return {
        key: program.id,
        team: program.name,
        datasets: program.datasets.map((dataset) => {
          return {
            id: dataset.id,
            title: dataset.name,
          };
        }),
      };
    })
  );
```

The end result is double square brackets array (array of array) instead of single square brackets as shown in this error: 

```
Argument of type '(program: {    key: string;    team: string;    datasets: Dataset[];}) => JSX.Element' is not assignable to parameter of type '(value: { key: string; team: string; datasets: { id: string; title: string; }[]; }[], index: number, array: { key: string; team: string; datasets: { id: string; title: string; }[]; }[][]) => Element'.
  Types of parameters 'program' and 'value' are incompatible.
    Type '{ key: string; team: string; datasets: { id: string; title: string; }[]; }[]' is missing the following properties from type '{ key: string; team: string; datasets: Dataset[]; }': key, team, datasetsts(2345)
```

seerocode replied:
NVM! I fixed it with a flat map.

jnu commented on client/src/components/DatasetDetails.tsx:
```diff
@@ -1,42 +1,67 @@
-  const params = useParams<RouteParams>();
+  const { datasetId } = useParams<RouteParams>();
+
+  const { data, loading, error } = useQuery<GetDataset, GetDatasetVariables>(
```
not necessary for this PR but would be good to have an error screen to display when something goes wrong

seerocode replied:
Probably would be good to make error handling a story so we can go through the different experiences for each component.

jnu commented on client/src/components/Home.tsx:
```diff
@@ -3,104 +3,120 @@ import { Tag, Button, Table, Space } from "antd";
   },
 ];
 
+const getTableData = (queryData: any) => {
```
don't we have the right type for this? would be nice to figure out what it is on this line, then the rest of the lines should be inferrable

seerocode replied:
Fixed!

jnu commented on client/src/components/Home.tsx:
```diff
@@ -3,104 +3,120 @@ import { Tag, Button, Table, Space } from "antd";
+        (dataset: { id: string; name: string; tags: any[] }) => {
+          rowData.push({
+            id: dataset.id,
+            team: "BBC News",
```
we should be able to get the Team Name from within the query, right?

seerocode replied:
Fixed!

jnu reviewed (commented):
this is great, @ro!!!! i'm apparently not able to "approve" since i opened the PR in the first place, but feel free to approve and merge.

Since this PR is getting big I'd prefer to focus on merging soon and then putting up subsequent PRs to iterate! thanks so much & great work!!

Commit:
fix(ch129): get program name dynamically

Commit:
fix(ch129): Fix sidebar gql query to a flat map

seerocode reviewed (approved):
I approve of my work 🤣
